### PR TITLE
don't calculate TDD or basal:bolus ratio in Basics if >= 1 day boluses missing

### PR DESCRIPTION
**Pull request by jebeck**
For your review @jh-bate 

LMK if you need some additional context beyond the Trello card for this. I'll add a couple of code comments in line for you too.

### CODE REVIEW

**Comment by jebeck** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;gemsReIJ&#x2F;17-ux-1-dev-2-tst-1-blip-basics-don-t-show-tdd-and-basal-bolus-ratio-under-certain-circumstances"> &lt;ux: 1, dev: 2, tst: 1&gt; [blip] [basics] Don&#39;t show TDD and basal:bolus ratio under certain circumstances</a>

File: plugins/blip/basics/logic/datamunger.js [hunk 1]
```diff
@@ -70,8 +70,23 @@ module.exports = function(bgClasses) {
...
+      var pastBolusDays = _.reject(
+        Object.keys(basicsData.data.bolus.dataByDate),
+        function(date) { return date === mostRecent; }
+      );
+      if (pastBolusDays.length < pastDays.length) {
+        return {
+          basalBolusRatio: null,
+          totalDailyDose: null,
+        };
+      }
```
Comment (jebeck) [inline plugins/blip/basics/logic/datamunger.js]: This is the most important logic change in this PR. The `dataByDate` object for each data type is created from a crossfilter grouping by date, so it only includes the days for which there are actually data. So I'm comparing the length of this (excepting the most recent day if there were boluses on it) against the length of the days (except for the most recent) in order to see if any days are missing boluses. I don't exactly think this is a _fragile_ strategy, but it's a little indirect, so I have some funny feelings about it. ¯\_(ツ)_/¯

File: plugins/blip/basics/components/chart/BasalBolusRatio.js [hunk 2]
```diff
@@ -62,26 +76,49 @@ var BasalBolusRatio = React.createClass({
...
+    var basal = _.get(data, ['basalBolusRatio', 'basal'], null);
+    var bolus = _.get(data, ['basalBolusRatio', 'bolus'], null);
+    var basalPercentClasses = cx({
+      'BasalBolusRatio-percent': true,
+      'BasalBolusRatio-percent--basal': !!basal,
+      'BasalBolusRatio--nodata': !basal
+    });
+    var basalLabelClasses = cx({
+      'BasalBolusRatio-label': true,
+      'BasalBolusRatio-label--basal': !!basal,
```
Comment (jh-bate) [inline plugins/blip/basics/components/chart/BasalBolusRatio.js]: just a syntax question `!!` ??

File: plugins/blip/basics/logic/datamunger.js [hunk 3]
```diff
@@ -70,8 +70,23 @@ module.exports = function(bgClasses) {
       return bgDistribution;
     },
     calculateBasalBolusStats: function(basicsData) {
-      var basals = basicsData.data.basal.data;
+      var pastDays = _.filter(basicsData.days, {type: 'past'});
+      // if one or more of the days (excepting most recent) don't have any boluses
+      // then don't calculate these stats at all, since may be inaccurate if
+      // long-running basals exist
+      var mostRecent = _.filter(basicsData.days, {type: 'mostRecent'})[0].date;
```
Comment (jh-bate) [inline plugins/blip/basics/logic/datamunger.js]: and there is no chance the first element wont exist?

File: plugins/blip/basics/logic/datamunger.js (hunk 1 above)
Comment (jh-bate) [inline plugins/blip/basics/logic/datamunger.js]: ¯\(ツ)/¯ at least there are no strange names :)

File: plugins/blip/basics/components/chart/BasalBolusRatio.js (hunk 2 above)
Comment (jebeck) [inline plugins/blip/basics/components/chart/BasalBolusRatio.js]: Yah, this is one of the accepted ways to coerce to a Boolean in JavaScript - `!!x` is the equivalent of `Boolean(x)` (but requires less typing 😜). I first learned it from Gordon, and @krystophv and I talked about it recently on one of his PRs, and I think we decided we like it generally. (@krystophv chime in if that's not the takeaway you remember from our exchange.)

File: plugins/blip/basics/logic/datamunger.js (hunk 3 above)
Comment (jebeck) [inline plugins/blip/basics/logic/datamunger.js]: It really shouldn't ever not exist, but I adjusted the code just in case...

**Comment by jh-bate** [comment]
LGTM @jebeck

**Pull request merged**